H.W. Redux

ark-web-dev opened the pull request:
GH-pages - https://ark-web-dev.github.io/Reviewer.bro/

### Перевёл приложение на Redux (без toolkit).

### Хотел бы изменить/добавить (возможно) в будущем:

- Рекомпозицию структуры приложения и некоторых компонентов. Некоторые компоненты нужно сделать более самостоятельными?

- Оптимизацию работы с localStorage. Пересмотреть некоторую логику. Стоит ли держать логику добавления в localStorage в async actions?

- С Redux логика в компонентах стала меньше. Возможно, стоит вынести её из кастомных хуков? (Хотя, нравится идея разделения логики и представления:)

ark-web-dev committed:
h.w. redux

aartem1 commented on src/store/entities/contributors/thunk/fetchContributorsThunk.ts:
```diff
@@ -0,0 +1,25 @@
+import { getGithubContributors } from '@/shared/API';
+
+export const fetchContributorsThunk =
+  (login: string, repoName: string) =>
```
в async action ты можешь получить доступ к данным стора без передачи их параметрами
Поправь, плиз, и в других местах где делаешь похожим образом

ark-web-dev replied:
Совсем забыл об этом когда преводил на Redux) Спасибо!
Поправил [ee76ada](https://github.com/ark-web-dev/Reviewer.bro/pull/3/commits/ee76ada420130ea2b4e36ae880fea00d26590b37)

aartem1 commented on src/store/entities/contributors/thunk/fetchContributorsThunk.ts:
```diff
@@ -0,0 +1,25 @@
+      return;
+    }
+
+    setLocalStorage('current-contributors', contributors);
```
я не уверен, что в localStorage стоит хранить что-либо кроме settings
Все остальное это динамические данные, которые могу менять и, кажется, нет смысла их хранить в двух местах

Суть же такая, мы сохраняем настройки, чтобы после перезагрузки страницы могли не заполнять данные заново. А то, что мы подгружаем, по идее должно заполнять при загрузке страницы на основе настроек (если они есть)

aartem1 replied:
Обычно мы не дублируем весь redux стор в localStorage
localStorage используется точечно для каких-то данных, которые должны сохраняться условно "на всегда"

ark-web-dev replied:
Хм, видно изначально не так понял этот момент. Я еще размышлял, мол если у меня запросы автоматически делаются, зачем их делать лишний раз, загрузим из localStorage)) а сейчас подумал, у нас же могут данные в localStorage устареть, например у юзера добавятся новые репозитории, или у старых появятся новые контрибьюторы... то есть, тем более лучше подгружать новые данные. Спасибо что указал на этот момент!

Поправил (полностью переписал логику взаимодействия с localStorage):
[90e0646](https://github.com/ark-web-dev/Reviewer.bro/pull/3/commits/90e06464315b753ca7d5f9875a5f2667ae4adbe5) 
[ec1c1b1](https://github.com/ark-web-dev/Reviewer.bro/pull/3/commits/ec1c1b1fc3d87422fc646dfc809b1a09f39c3d2a) 
[d8244c4](https://github.com/ark-web-dev/Reviewer.bro/pull/3/commits/d8244c42a17c19bb882c5303b786299b704a0184)

aartem1 commented:
> С Redux логика в компонентах стала меньше. Возможно, стоит вынести её из кастомных хуков? (Хотя, нравится идея разделения логики и представления:)

Да, мне тоже понравилась такая организация структуры

aartem1 commented:
> Оптимизацию работы с localStorage. Пересмотреть некоторую логику. Стоит ли держать логику добавления в localStorage в async actions?

стоит сократить количество данных в локалстораж

Оставить там только самое необходимое

aartem1 commented:
При вводе данных вот сюда
<img width="684" alt="Снимок экрана 2024-02-12 в 11 16 57" src="https://github.com/ark-web-dev/Reviewer.bro/assets/26367363/ff596d84-6f76-410f-a4d5-36c32b664584">
у тебя запускается подгрузка данных о пользователе, но нет никакой индикации этого процесса, это выглядит немного странно для пользователя, когда я начинаю что-то вводить, экран начинает просто дергаться и меняться

Это минорное замечание, править не обязательно, но просто обратил внимание

ark-web-dev committed:
refactor: getting data for asynс actions, from the store, instead of function parameters

ark-web-dev committed:
update: only necessary settings are stored in localStorage

the logic of storing and retrieving from the localStorage has been completely changed

ark-web-dev committed:
fix: getting currentRepo from store instead of useState

ark-web-dev committed:
fix: added missing fetch loading

ark-web-dev committed:
update: elements loading with animation
